module/platform/Webots/src/Webots.cpp

```diff
@@ -0,0 +1,656 @@
+            // TODO(Cameron) Try to reconnect after a delay.
+            // setup_connection(server_address, port);
+            log<NUClear::FATAL>("Quitting due to failed connection attempt");
+            powerplant.shutdown();
+            return;
+        }
+
+        // Initaliase the string with ???????
+        std::string initial_message = std::string(7, '?');
+        const int n                 = recv(fd, initial_message.data(), sizeof(initial_message), 0);
+
+        if (n >= 0) {
+            if (initial_message == "Welcome") {
+                // good
+                log<NUClear::INFO>(fmt::format("Connected to {}:{}", server_address, port));
+            }
+            else if (initial_message == "Refused") {
+                log<NUClear::FATAL>(
+                    fmt::format("Connection to {}:{} refused: your ip is not white listed.", server_address, port));
+                // Halt and don't retry as reconnection is pointless.
+                close(fd);
+                powerplant.shutdown();
+            }
+            else {
+                log<NUClear::FATAL>(fmt::format("{}:{} sent unknown initial message", server_address, port));
+                // Halt and don't retry as the other end is clearly not Webots
+                close(fd);
+                powerplant.shutdown();
+            }
+        }
+        else {
+            // There was nothing sent
+            log<NUClear::DEBUG>("Connection was closed.");
+            setup_connection(server_address, port);
+            return;
+        }
+
+        // Set the real time of the connection initiation
+        connect_time = NUClear::clock::now();
+        // Reset the simulation connection time
+        utility::clock::last_update = std::chrono::steady_clock::now();
+
+        // Now that we are connected, we can set up our reaction handles with this file descriptor
+
+        // Receiving
+        read_io = on<IO>(fd, IO::READ).then([this]() {
+            // ************************** Receiving ***************************************
+            // Get the size of the message
+            uint32_t Nn;
+            if (recv(fd, &Nn, sizeof(Nn), 0) != sizeof(Nn)) {
+                log<NUClear::ERROR>("Failed to read message size from TCP connection");
+                return;
+            }
+
+            // Convert from network endian to host endian
+            const uint32_t Nh = ntohl(Nn);
+
+            // Get the message
+            std::vector<char> data(Nh, 0);
+            uint64_t message = uint64_t(recv(fd, data.data(), Nh, MSG_WAITALL));
+            if (message != Nh) {
+                log<NUClear::ERROR>("Failed to read message from TCP connection");
+                return;
+            }
+
+            log<NUClear::TRACE>("Received sensor measurements");
+
+            // Deserialise the message into a neutron
+            SensorMeasurements msg = NUClear::util::serialise::Serialise<SensorMeasurements>::deserialise(data);
```

Comment:
If `Nh == 0` we should set `msg` to a default constructed `SensorMeasurements` message (if I am not mistaken, a protobuf message with all fields set to their defaults has a 0 byte serialised length)